src/ril_binder_radio.c

```diff
@@ -2338,6 +2606,62 @@ ril_binder_radio_decode_signal_strength_1_2(
     return FALSE;
 }
 
+/**
+ * @param signalStrength SignalStrength information as defined in types.hal
+ */
+static
+gboolean
+ril_binder_radio_decode_signal_strength_1_4(
+    GBinderReader* in,
+    GByteArray* out)
+{
+    const RadioSignalStrength_1_4* strength = gbinder_reader_read_hidl_struct
+        (in, RadioSignalStrength_1_4);
+
+    if (strength) {
+        const RadioSignalStrengthGsm* gsm = &strength->gsm;
+        const RadioSignalStrengthWcdma* wcdma = &strength->wcdma.base;
+
+        /* GW_SignalStrength */
+        if (wcdma->signalStrength <= 31 && gsm->signalStrength > 31) {
+            /*
+             * Presumably, 3G signal. The wcdma field did't exist in RIL
+             * socket times.
+             *
+             * Valid signal strength values for both 2G and 3G are (0-31, 99)
+             * as defined in TS 27.007 8.5
+             */
+            grilio_encode_int32(out, wcdma->signalStrength);
+            grilio_encode_int32(out, wcdma->bitErrorRate);
+        } else {
+            grilio_encode_int32(out, gsm->signalStrength);
+            grilio_encode_int32(out, gsm->bitErrorRate);
+        }
+
+        /* CDMA_SignalStrength */
+        grilio_encode_int32(out, strength->cdma.dbm);
+        grilio_encode_int32(out, strength->cdma.ecio);
+
+        /* EVDO_SignalStrength */
+        grilio_encode_int32(out, strength->evdo.dbm);
+        grilio_encode_int32(out, strength->evdo.ecio);
+        grilio_encode_int32(out, strength->evdo.signalNoiseRatio);
+
+        /* LTE_SignalStrength_v8 */
+        grilio_encode_int32(out, strength->lte.signalStrength);
+        grilio_encode_int32(out, strength->lte.rsrp);
+        grilio_encode_int32(out, strength->lte.rsrq);
+        grilio_encode_int32(out, strength->lte.rssnr);
+        grilio_encode_int32(out, strength->lte.cqi);
+        grilio_encode_int32(out, strength->lte.timingAdvance);
+
+        /* TD_SCDMA_SignalStrength */
+        grilio_encode_int32(out, strength->tdscdma.rscp);
+        return TRUE;
+    }
+    return FALSE;
+}
```

Comment:
The above looks suspiciously similar to `ril_binder_radio_decode_signal_strength_1_2()` and even `ril_binder_radio_decode_signal_strength()`. Consider moving common code to one function. e.g.
```c
static
void
ril_binder_radio_decode_signal_strength_common(
    const RadioSignalStrengthGsm* gsm,
    const RadioSignalStrengthCdma* cdma,
    const RadioSignalStrengthEvdo* evdo,
    const RadioSignalStrengthLte* lte,
    const RadioSignalStrengthTdScdma* tdScdma,
    const RadioSignalStrengthWcdma* wcdma,
    GByteArray* out)
{
    /* GW_SignalStrength */
    if (wcdma && wcdma->signalStrength <= 31 && gsm->signalStrength > 31) {
        /*
         * Presumably, 3G signal. The wcdma field did't exist in RIL
         * socket times.
         *
         * Valid signal strength values for both 2G and 3G are (0-31, 99)
         * as defined in TS 27.007 8.5
         */
        grilio_encode_int32(out, wcdma->signalStrength);
        grilio_encode_int32(out, wcdma->bitErrorRate);
    } else {
        grilio_encode_int32(out, gsm->signalStrength);
        grilio_encode_int32(out, gsm->bitErrorRate);
    }

    /* CDMA_SignalStrength */
    grilio_encode_int32(out, cdma->dbm);
    grilio_encode_int32(out, cdma->ecio);

    /* EVDO_SignalStrength */
    grilio_encode_int32(out, evdo->dbm);
    grilio_encode_int32(out, evdo->ecio);
    grilio_encode_int32(out, evdo->signalNoiseRatio);

    /* LTE_SignalStrength_v8 */
    grilio_encode_int32(out, lte->signalStrength);
    grilio_encode_int32(out, lte->rsrp);
    grilio_encode_int32(out, lte->rsrq);
    grilio_encode_int32(out, lte->rssnr);
    grilio_encode_int32(out, lte->cqi);
    grilio_encode_int32(out, lte->timingAdvance);

    /* TD_SCDMA_SignalStrength */
    grilio_encode_int32(out, tdScdma->rscp);
}
```
and calling it like this:
```c
static
gboolean
ril_binder_radio_decode_signal_strength_1_4(
    GBinderReader* in,
    GByteArray* out)
{
    const RadioSignalStrength_1_4* strength = gbinder_reader_read_hidl_struct
        (in, RadioSignalStrength_1_4);

    if (strength) {
        RadioSignalStrengthTdScdma tdscdma;

        tdscdma.rscp = strength->tdscdma.rscp;
        ril_binder_radio_decode_signal_strength_common(&strength->gsm,
            &strength->cdma, &strength->evdo, &strength->lte,
            &tdscdma, &strength->wcdma.base, out);
        return TRUE;
    }
    return FALSE;
}
```
from all 3 places. Copy/paste is evil.